### PR TITLE
Add secure mode

### PR DESCRIPTION
**Pull request by alexandresablayrolles**
PrivacyEngine creates the random_number_generator, that is passed on to DPOptimizer and DPDataLoader (that passes it to poisson sampler).

For deterministic run, the user can call privacy_engine.set_seed() before make_private()

### CODE REVIEW

File: opacus/privacy_engine.py [hunk 1]
```diff
@@ -31,7 +33,25 @@ def forbid_accumulation_hook(module: nn.Module, _):
...
+            except ImportError as e:
+                msg = (
+                    "To use secure RNG, you must install the torchcsprng package! "
+                    "Check out the instructions here: https://github.com/pytorch/csprng#installation"
+                )
+                raise ImportError(msg) from e
+
+            self.secure_rng = csprng.create_random_device_generator("/dev/urandom")
+        else:
+            warnings.warn(
```
[the remark this answers is not in the document]
Comment (karthikprasad) [reply opacus/privacy_engine.py:50]: TIL. I always considered `warning` to be more of an internal program "problem" that can be fixed by user (such as deprecation, future proofing, etc.) whereas `logging.warn` as more of non-problem that is acceptable, but not desirable.
The documentation makes it clearer and I agree with you and Alex that warnings.warn sounds like the right way do it.

That said, since we explicitly allow users to not secure mode via a flag, it is best to document that we will throw a warning when secure mode is off. To a new user, it will be baffling if the freshly baked opacus with default settings throws warnings.

[the remark this answers is not in the document]
Comment (ffuuugor) [reply opacus/privacy_engine.py:50]: > We should def. nudge users towards using secure_mode and this is a useful reminder

I'd argue that it highly depends on the use case. If you're doing research, you probably don't care and it's just annoying.
But since warnings are mostly treated the same as logged warnings, I agree it's ok to keep it as wanings. (and if your app is big enough that you want to do logger config, you should probably also consider using secure_mode)

File: opacus/privacy_engine.py [hunk 2]
```diff
@@ -128,7 +159,11 @@ def make_private_per_layer(
             batch_first,
             loss_reduction,
         )
-        data_loader = self._prepare_data_loader(data_loader, distributed=distributed)
+        if poisson_sampling:
+            data_loader = self._prepare_data_loader(
```
[the remark this answers is not in the document]
Comment (ffuuugor) [reply opacus/privacy_engine.py:163]: > The DataLoader can create a PoissonSampler or a regular sampler with the provided generator.

"the regular sampler" can actually be a lot of different things. We should probably respect the way users have configured their data loader and not change it without a good reason. And there're a lot of ways to configure sampling in the vanilla DataLoader: by providing sampler, batch_sampler or just a `shuffle` flag and a generator

File: opacus/utils/uniform_sampler.py [hunk 3]
```diff
@@ -73,7 +73,7 @@ def __init__(
         total_size: int,
         sample_rate: float,
         shuffle: bool = True,
-        seed: int = 0,
+        shuffle_seed: int = 0,
```
[the remark this answers is not in the document]
Comment (ffuuugor) [reply opacus/utils/uniform_sampler.py:71]: Hmm, after closer inspection it looks fine

`shuffle_seed` is used to control the data split between workers, not the data shuffling within one worker.

Client can control shuffling by passing `generator` to the data loader - when not in secure_mode, we reuse generator provided by the original data_loader.

One interesting question though is whether we want to keep the original `shuffle_seed` from the original data loader if it's using DistributedSampler - I'm not sure what's the right approach here

File: opacus/privacy_engine.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ffuuugor) [reply opacus/privacy_engine.py]: Updated this part: secure shuffling is now independent from poisson sampling

File: opacus/privacy_engine.py [hunk 4]
```diff
@@ -128,7 +162,13 @@ def make_private_per_layer(
             batch_first,
             loss_reduction,
         )
```
Comment (karthikprasad) [inline opacus/privacy_engine.py:164]: Looks like a bug, could you please delete the second call to _prepare_model?

File: opacus/privacy_engine.py [hunk 5]
```diff
@@ -113,11 +143,15 @@ def make_private_per_layer(
...
         loss_reduction: str = "mean",
+        noise_seed: Optional[int] = None,
         poisson_sampling: bool = True,
         try_fix_incompatible_modules: bool = False,
     ):
         distributed = type(module) is DPDDP
 
+        if noise_seed and self.secure_mode:
+            raise ValueError("Passing seed is prohibited in secure mode")
+
```
Comment (karthikprasad) [inline opacus/privacy_engine.py:154]: Should we move this to `validate()` ?

File: opacus/utils/uniform_sampler.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (karthikprasad) [reply opacus/utils/uniform_sampler.py:71]: I'm still not seeing how this specific arg is set/controlled by the generator. Will it ever be non-zero? I'll see again after a cup of coffee.:P

**Review by karthikprasad — APPROVED** [review]
Added some comments, mostly nits.

File: examples/cifar10.py [hunk 6]
```diff
@@ -371,7 +370,9 @@ def main():
             for param_group in optimizer.param_groups:
                 param_group["lr"] = lr
 
-        train_duration = train(args, model, train_loader, optimizer, privacy_engine, epoch, device)
+        train_duration = train(
+            args, model, train_loader, optimizer, privacy_engine, epoch, device
+        )
         top1_acc = test(args, model, test_loader, device)
 
         # remember best acc@1 and save checkpoint
```
[the remark this answers is not in the document]
Comment (ffuuugor) [reply examples/cifar10.py:378]: > Github doesn't allow you to add comment on unchanged files? WTH?!!

I know, right?!

[the remark this answers is not in the document]
Comment (ffuuugor) [reply examples/cifar10.py:378]: tbh, in the context of tutorials secure_rng warnings looks especially inappropriate

File: opacus/privacy_engine.py (hunk 5 above)
Comment (ffuuugor) [reply opacus/privacy_engine.py:154]: You mean crate new method?

File: opacus/utils/uniform_sampler.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ffuuugor) [reply opacus/utils/uniform_sampler.py:71]: a) It's not controlled by a generator, nor it's supposed to.

There are two distinct elements of randomness with DistributedSampler: how to split the dataset between workers and how to select the items from the dataset.
The latter (controlled by shuffle seed) is not really that important: it affects neither model performance nor privacy guarantee.
The former is important, and hence we use user-provided generator (secure rng in secure mode) to create a dataset mask

b) Will it ever be non-zero?
Not in the default setting. 

Having `shuffle_seed` as a parameter is important for a self-sufficient Sampler. If we consider this sampler as a individual class, you do want to have the option to control the splitting between workers (after all, that's what `torch.utils.data.distributed.DistributedSampler` does).
On the other hand, I don't see a good case for adding this as an additional parameter to `DPDataLoader` or `PrivacyEngine`. We don't really expose every downstream dependency parameter to the user - that would be overwhelming. My thinking is that if this level of precision is necessary, people can construct their own `DPDataLoader` and pass custom `DistributedUniformWithReplacementSampler`
